### PR TITLE
Draft plumbing for test vector generation

### PR DESCRIPTION
**Pull request by chris-wood**
Some questions to consider:

1. Does the "test then spit out vector" pattern make sense?
2. Should the private key marshal interface be defined elsewhere? 
3. Should we add a method to the cipher suite interface that returns true/false for whether or not test vectors are supported? (They'd be true for all except SIKE, since we don't have a standard way to encode those private keys [that I'm aware of].)

### CODE REVIEW

**Comment by bifurcation** [comment]
Couple of comments here.

0. We should try to be more complete, i.e., capture all computed values so that another implementor can compare at any point.  We should also test the behavior of the multi-shot interface (i.e., that the counter / masking work properly).  

0.1. that will mean the JSON struct should be something like:

```
type TestCase struct {
	// Crypto parameters
	mode Mode
	kem  KEMID
	kdf  KDFID
	aead AEADID

	// Private keys
	skRm, skEm []byte

	// Key schedule inputs
	pkRm, zz, enc, info, psk, pskID, pkIm  []byte

	// Values computed in KeySchedule
	ciphersuite, pskID_hash, info_hash, context, secret, key, nonce []byte

	// Encrypted values at various offsets
	pt0, aad0, ct0       []byte
	pt1, aad1, ct1       []byte
	pt256, aad256, ct256 []byte
}
```

0.2. You'll need a way to inject the preset ephemeral key.  I would propose making a package-private (lower-case) method on the DHKEM interface that lets you set it into a private field in the DHKEM object, and then having Encap use it when set.  Then you can make a Ciphersuite, set the ephemeral private key, and everything else should work as normal.

0.3. You'll need a way to collect the intermediate results from KeySchedule.  I would propose making a package-private struct to hold these values, having private fields on EncryptContext/DecryptContext of that type, and having `encryptionContext` populate it with the results.  (And while you're there, rename `encryptionContext` to `keySchedule()`!)

1. We should keep the test vector generation separate from the actual testing.  As long as they both run in a given instance, we'll know they're both correct.  (Actually, we can also run the tests on the generated vectors.)  And we'll want to do some different things for the test vectors vs. the real tests.

2. It's find to have the marshal-privkey interfaces here.

3. No, we should just have a list of cases that we want to generate test vectors for, which will be less than the full combinatorial explosion.

File: crypto.go [hunk 1]
```diff
@@ -30,10 +30,13 @@ type dhScheme interface {
 	Unmarshal(enc []byte) (KEMPublicKey, error)
 	DH(priv KEMPrivateKey, pub KEMPublicKey) ([]byte, error)
 	PublicKeySize() int
+	setEphemeralKeyPair(sk KEMPrivateKey, pk KEMPublicKey)
 }
 
 type dhkemScheme struct {
 	group dhScheme
+	skE KEMPrivateKey
```
Comment (bifurcation) [inline crypto.go]: It would be helpful to add a comment explaining these, since it might not be totally obvious.

File: crypto.go [hunk 2]
```diff
@@ -646,10 +684,10 @@ const (
 )
 
 var kems = map[KEMID]KEMScheme{
-	DHKEM_X25519: dhkemScheme{x25519Scheme{}},
-	DHKEM_X448:   dhkemScheme{x448Scheme{}},
-	DHKEM_P256:   dhkemScheme{ecdhScheme{curve: elliptic.P256()}},
-	DHKEM_P521:   dhkemScheme{ecdhScheme{curve: elliptic.P521()}},
+	DHKEM_X25519: dhkemScheme{group: x25519Scheme{}, skE: nil, pkE: nil},
```
Comment (bifurcation) [inline crypto.go]: No need to specify those fields; by default they're set to their zero value, which is nil.

File: crypto_test.go [hunk 3]
```diff
@@ -17,10 +17,10 @@ func randomBytes(size int) []byte {
 
 func TestKEMSchemes(t *testing.T) {
 	schemes := []KEMScheme{
-		dhkemScheme{x25519Scheme{}},
-		dhkemScheme{x448Scheme{}},
-		dhkemScheme{ecdhScheme{curve: elliptic.P256()}},
-		dhkemScheme{ecdhScheme{curve: elliptic.P521()}},
+		dhkemScheme{group: x25519Scheme{}, skE: nil, pkE: nil},
```
Comment (bifurcation) [inline crypto_test.go]: Ditto down here.

File: hpke.go [hunk 4]
```diff
@@ -136,7 +136,14 @@ type hpkeContext struct {
 	infoHash  []byte `tls:"head=none"`
 }
 
-func keySchedule(suite CipherSuite, mode HPKEMode, pkR KEMPublicKey, zz, enc, info, psk, pskID, pkIm []byte) (key, nonce []byte, err error) {
+type contextParameters struct {
+	context hpkeContext
```
Comment (bifurcation) [inline hpke.go]: I think you need at least one more field here, for the marshaled `context` value.  Maybe also the `keyContext` and `nonceContext` values.

File: hpke.go [hunk 5]
```diff
@@ -136,7 +136,14 @@ type hpkeContext struct {
...
+	secret []byte
+	key []byte
+	nonce []byte
+}
+
+func keySchedule(suite CipherSuite, mode HPKEMode, pkR KEMPublicKey, zz, enc, info, psk, pskID, pkIm []byte) (params contextParameters, err error) {
```
Comment (bifurcation) [inline hpke.go]: Is there a specific reason you're using named return values here?  Generally, my impression is that they're to be avoided in all but very small functions.

File: crypto.go (hunk 1 above)
Comment (chris-wood) [reply crypto.go]: Yep, will do!

File: crypto.go (hunk 2 above)
Comment (chris-wood) [reply crypto.go]: Oh, cool. I'll remove them!

File: hpke.go (hunk 5 above)
Comment (chris-wood) [reply hpke.go]: No reason other than I don't know how to write idiomatic Go code. :-)

File: crypto.go [hunk 6]
```diff
@@ -47,22 +57,61 @@ func (s dhkemScheme) Marshal(pk KEMPublicKey) []byte {
...
+func (s dhkemScheme) ephemeralPrivate() KEMPrivateKey {
+	return s.skE
+}
+
+func (s dhkemScheme) getEphemeralKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey, error) {
+	var skE KEMPrivateKey
+	var pkE KEMPublicKey
+	var err error
+
+	if s.skE != nil && s.pkE != nil {
```
Comment (bifurcation) [inline crypto.go]: You could streamline this method by just:

```
if (...) {
  return s.skE, s.pkE, nil
}

/* generate them */
```

File: crypto.go [hunk 7]
```diff
@@ -47,22 +57,61 @@ func (s dhkemScheme) Marshal(pk KEMPublicKey) []byte {
...
+		skE, pkE, err = s.group.GenerateKeyPair(rand)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return skE, pkE, nil
+}
+
+func (s dhkemScheme) Encap(rand io.Reader, pkR KEMPublicKey) ([]byte, []byte, KEMPrivateKey, KEMPublicKey, error) {
```
Comment (bifurcation) [inline crypto.go]: You don't need to return the public key here, since you're already returning it in marshaled form.

File: crypto.go [hunk 8]
```diff
@@ -79,25 +128,24 @@ func (s dhkemScheme) Decap(enc []byte, skR KEMPrivateKey) ([]byte, error) {
 	return zz, nil
 }
 
-func (s dhkemScheme) AuthEncap(rand io.Reader, pkR KEMPublicKey, skI KEMPrivateKey) ([]byte, []byte, error) {
-
-	skE, pkE, err := s.group.GenerateKeyPair(rand)
+func (s dhkemScheme) AuthEncap(rand io.Reader, pkR KEMPublicKey, skI KEMPrivateKey) ([]byte, []byte, KEMPrivateKey, KEMPublicKey, error) {
```
Comment (bifurcation) [inline crypto.go]: Ditto here

File: crypto.go [hunk 9]
```diff
@@ -167,10 +217,29 @@ func (s ecdhScheme) GenerateKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey
...
 
+func (s ecdhScheme) MarshalPrivate(sk KEMPrivateKey) []byte {
+	if sk == nil {
+		return nil
+	}
+
+	raw := sk.(*ecdhPrivateKey)
+	encodedPoint := elliptic.Marshal(raw.curve, raw.x, raw.y)
+
+	bundle := make([]byte, 2)
```
Comment (bifurcation) [inline crypto.go]: No need, technically, to encode the point here, and if you get rid of that, you don't need the length.  That is, you could just `return raw.d.Bytes()`.

File: crypto.go [hunk 10]
```diff
@@ -180,6 +249,30 @@ func (s ecdhScheme) Unmarshal(enc []byte) (KEMPublicKey, error) {
 	return &ecdhPublicKey{s.curve, x, y}, nil
 }
 
+func (s ecdhScheme) UnmarshalPrivate(enc []byte) (KEMPrivateKey, error) {
```
Comment (bifurcation) [inline crypto.go]: Parallel comment here.

File: crypto.go [hunk 11]
```diff
@@ -26,13 +28,21 @@ type dhScheme interface {
 	ID() KEMID
 	GenerateKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey, error)
 	Marshal(pk KEMPublicKey) []byte
+	MarshalPrivate(sk KEMPrivateKey) []byte
 	Unmarshal(enc []byte) (KEMPublicKey, error)
+	UnmarshalPrivate(enc []byte) (KEMPrivateKey, error)
 	DH(priv KEMPrivateKey, pub KEMPublicKey) ([]byte, error)
 	PublicKeySize() int
+
+	setEphemeralKeyPair(sk KEMPrivateKey, pk KEMPublicKey)
```
Comment (bifurcation) [inline crypto.go]: I would just take the PrivateKey here (and only store that).  It avoids the risk of mismatch, and means you only have to store one thing.

Comment (bifurcation) [reply crypto.go]: Also, I don't see any calls to this method?

File: crypto.go [hunk 12]
```diff
@@ -26,13 +28,21 @@ type dhScheme interface {
...
+	ephemeralPublic() KEMPublicKey
+	ephemeralPrivate() KEMPrivateKey
```
Comment (bifurcation) [inline crypto.go]: Are these needed?  Don't see any calls, and they seem duplicative of `getEphemeralKeyPair`

File: crypto.go [hunk 13]
```diff
@@ -391,12 +566,20 @@ func (s sikeScheme) GenerateKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey
...
+		return nil
+	}
 	raw := pk.(*sikePublicKey)
 	out := make([]byte, raw.pub.Size())
 	raw.pub.Export(out)
 	return out
 }
 
+func (s sikeScheme) MarshalPrivate(sk KEMPrivateKey) []byte {
+	// There's no standard way to encode a SIKE private key, so return nil (for now).
```
Comment (bifurcation) [inline crypto.go]: Don't really care about "standard" here, since we could just specify it in the document.  More salient that we we're not going to generate test vectors right now.

Comment (bifurcation) [reply crypto.go]: I wonder if these unimplemented methods should panic instead of returning.  Fail fast.

File: crypto.go [hunk 14]
```diff
@@ -641,8 +843,26 @@ var aeads = map[AEADID]AEADScheme{
 	AEAD_CHACHA20POLY1305: chachaPolyScheme{},
 }
 
-func AssembleCipherSuite(kemID KEMID, kdfID KDFID, aeadID AEADID) (CipherSuite, error) {
-	kem, ok := kems[kemID]
+func AssembleKEMWithState(kemID KEMID, sk KEMPrivateKey, pk KEMPublicKey) (KEMScheme, bool) {
```
Comment (bifurcation) [inline crypto.go]: Please `go fmt -s`

Comment (bifurcation) [reply crypto.go]: This method should be module-private.

File: crypto.go [hunk 15]
```diff
@@ -641,8 +843,26 @@ var aeads = map[AEADID]AEADScheme{
...
+		case DHKEM_X448: return dhkemScheme{group: x448Scheme{}, skE: sk, pkE: pk}, true
+		case DHKEM_P256: return dhkemScheme{group: ecdhScheme{curve: elliptic.P256()}, skE: sk, pkE: pk}, true
+		case DHKEM_P521: return dhkemScheme{group: ecdhScheme{curve: elliptic.P521()}, skE: sk, pkE: pk}, true
+		case KEM_SIKE503: return sikeScheme{field: sidh.Fp503}, true
+		case KEM_SIKE751: return sikeScheme{field: sidh.Fp751}, true
+	}
+	return nil, false
+}
+
+func AssembleCipherSuite(kemID KEMID, kdfID KDFID, aeadID AEADID, sk KEMPrivateKey, pk KEMPublicKey) (CipherSuite, error) {
```
Comment (bifurcation) [inline crypto.go]: I would revert this method, and make a duplicate, module-private one if you want.  (Or just repurpose the above method so that it assembles the whole suite instead of just the KEM.  We shouldn't be exposing our test points to the world.

File: crypto_test.go (hunk 3 above)
Comment (bifurcation) [inline crypto_test.go]: Nil args not necessary.

File: hpke.go [hunk 16]
```diff
@@ -79,6 +90,48 @@ func logVal(name string, value []byte) {
 	}
 }
 
+///////
+// Exported test vector structure
+type EncryptionTestVector struct {
```
Comment (bifurcation) [inline hpke.go]: Strong preference to keep this inside the module boundary.

File: hpke.go [hunk 17]
```diff
@@ -106,13 +159,13 @@ func verifyMode(suite CipherSuite, mode HPKEMode, psk, pskID, pkIm []byte) error
 
 	ok := false
 	switch mode {
-	case modeBase:
+	case ModeBase:
```
Comment (bifurcation) [inline hpke.go]: If we keep the tests inside the module boundary, no need for these constants to be public.

File: hpke.go [hunk 18]
```diff
@@ -135,10 +188,24 @@ type hpkeContext struct {
 	infoHash  []byte `tls:"head=none"`
 }
 
-func keySchedule(suite CipherSuite, mode HPKEMode, pkR KEMPublicKey, zz, enc, info, psk, pskID, pkIm []byte) (key, nonce []byte, err error) {
-	err = verifyMode(suite, mode, psk, pskID, pkIm)
+type ContextParameters struct {
+	Context []byte
```
Comment (bifurcation) [inline hpke.go]: I'm slightly tempted to capture the intermediate products that go into the context string, but I'll grant that you could pull them out of this string.

Comment (bifurcation) [reply hpke.go]: I'm a little split on whether to include the keyContext and nonceContext values.  There's a risk of people including null terminators, for example.  But I guess that's just one bit.  Maybe we should check to make sure the spec is clear.

File: hpke.go [hunk 19]
```diff
@@ -206,18 +283,24 @@ func (ctx *EncryptContext) Seal(aad, pt []byte) []byte {
 	return ct
 }
 
+func (ctx EncryptContext) Parameters() (SetupParameters, ContextParameters) {
```
Comment (bifurcation) [inline hpke.go]: Should be module-private

File: hpke.go [hunk 20]
```diff
@@ -230,24 +313,37 @@ func (ctx *DecryptContext) Open(aad, ct []byte) ([]byte, error) {
...
 
 func SetupBaseI(suite CipherSuite, rand io.Reader, pkR KEMPublicKey, info []byte) ([]byte, *EncryptContext, error) {
 	// zz, enc = Encap(pkR)
-	zz, enc, err := suite.KEM.Encap(rand, pkR)
+	zz, enc, skE, pkE, err := suite.KEM.Encap(rand, pkR)
 	if err != nil {
 		return nil, nil, err
 	}
 
+	// fmt.Printf("%x\n", zz)
```
Comment (bifurcation) [inline hpke.go]: Stray printf

File: hpke.go [hunk 21]
```diff
@@ -259,30 +355,30 @@ func SetupBaseR(suite CipherSuite, skR KEMPrivateKey, enc, info []byte) (*Decryp
 	}
 
 	pkR := skR.PublicKey()
-	key, nonce, err := keySchedule(suite, modeBase, pkR, zz, enc, info, defaultPSK(suite), defaultPSKID(suite), defaultPKIm(suite))
+	decryptionContextParams, err := keySchedule(suite, ModeBase, pkR, zz, enc, info, defaultPSK(suite), defaultPSKID(suite), defaultPKIm(suite))
```
Comment (bifurcation) [inline hpke.go]: This could probably just be `params` given the context.

File: hpke.go [hunk 22]
```diff
@@ -259,30 +355,30 @@ func SetupBaseR(suite CipherSuite, skR KEMPrivateKey, enc, info []byte) (*Decryp
...
 func SetupPSKI(suite CipherSuite, rand io.Reader, pkR KEMPublicKey, psk, pskID, info []byte) ([]byte, *EncryptContext, error) {
 	// zz, enc = Encap(pkR)
-	zz, enc, err := suite.KEM.Encap(rand, pkR)
+	zz, enc, _, _, err := suite.KEM.Encap(rand, pkR)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	key, nonce, err := keySchedule(suite, modePSK, pkR, zz, enc, info, psk, pskID, defaultPKIm(suite))
+	encryptionContextParams, err := keySchedule(suite, ModePSK, pkR, zz, enc, info, psk, pskID, defaultPKIm(suite))
```
Comment (bifurcation) [inline hpke.go]: Likewise, `params`

File: hpke.go [hunk 23]
```diff
@@ -259,30 +355,30 @@ func SetupBaseR(suite CipherSuite, skR KEMPrivateKey, enc, info []byte) (*Decryp
...
 
 	pkR := skR.PublicKey()
-	key, nonce, err := keySchedule(suite, modeBase, pkR, zz, enc, info, defaultPSK(suite), defaultPSKID(suite), defaultPKIm(suite))
+	decryptionContextParams, err := keySchedule(suite, ModeBase, pkR, zz, enc, info, defaultPSK(suite), defaultPSKID(suite), defaultPKIm(suite))
 	if err != nil {
 		return nil, err
 	}
 
-	return newDecContext(suite, key, nonce)
+	return newDecContext(suite, SetupParameters{}, decryptionContextParams)
```
Comment (bifurcation) [inline hpke.go]: Don't you want to be passing SetupParams here?  Also below.

File: hpke_test.go [hunk 24]
```diff
@@ -12,25 +14,23 @@ var (
...
 )
 
 func roundTrip(t *testing.T, kemID KEMID, kdfID KDFID, aeadID AEADID, enc *EncryptContext, dec *DecryptContext) {
-	for range make([]struct{}, rtts) {
-		encrypted := enc.Seal(aad, original)
-		decrypted, err := dec.Open(aad, encrypted)
-		if err != nil {
-			t.Fatalf("[%x, %x, %x] Error in Open: %s", kemID, kdfID, aeadID, err)
-		}
+	encrypted := enc.Seal(aad, original)
```
Comment (bifurcation) [inline hpke_test.go]: Why did you remove the multi-RTT version?

File: hpke.go (hunk 23 above)
Comment (chris-wood) [reply hpke.go]: Yep, just didn't get there yet!

File: hpke_test.go (hunk 24 above)
Comment (chris-wood) [reply hpke_test.go]: That's a leftover change I forgot to revert. I'll bring it back. (Previously when the vector generation was woven through the tests this multi-RT variant complicated things, so I dropped it.)

File: crypto.go (hunk 15 above)
Comment (chris-wood) [reply crypto.go]: Yep, I'll revert and make the ephemeral-setter one module-private.

File: crypto.go (hunk 13 above)
Comment (chris-wood) [reply crypto.go]: Yep, let's panic for now.

File: crypto.go (hunk 9 above)
Comment (chris-wood) [reply crypto.go]: Good point -- we can just recompute `(x,y)` from `d` when unmarshalling.

File: crypto.go (hunk 12 above)
Comment (chris-wood) [reply crypto.go]: Probably not!

File: crypto.go (hunk 11 above)
Comment (chris-wood) [reply crypto.go]: Yep, removed!

File: hpke_test.go [hunk 25]
```diff
@@ -155,3 +159,347 @@ func TestModes(t *testing.T) {
...
+			t.Fatalf("[%x, %x, %x] Incorrect encryption: [%x] != [%x]", vector.KEMID, vector.KDFID, vector.AEADID, dataCiphertext, encrypted)
+		}
+
+		if !bytes.Equal(decrypted, dataPlaintext) {
+			t.Fatalf("[%x, %x, %x] Incorrect decryption: [%x] != [%x]", vector.KEMID, vector.KDFID, vector.AEADID, decrypted, dataPlaintext)
+		}
+	}
+}
+
+func processTestVector(t *testing.T, vectorString string) {
```
Comment (bifurcation) [inline hpke_test.go]: It seems like you should have initiator and responder sides to this.  The Initiator side will call `SetupXI` and the responder side will call `SetupXR`.

In order to support that, I expect you'll want to pull out all the unmarshal logic below into `MarshalJSON` and `UnmarshalJSON` functions for the `testVector` struct, and have the fields in that struct actually just be the real `KEMPrivateKey`, `KEMPublicKey`, `[]byte`, etc. values.

File: hpke_test.go [hunk 26]
```diff
@@ -155,3 +159,347 @@ func TestModes(t *testing.T) {
...
+	var skR, skE KEMPrivateKey
+	var pkR, pkE KEMPublicKey
+
+	suite, err := AssembleCipherSuite(vector.KEMID, vector.KDFID, vector.AEADID)
+	if err != nil {
+		t.Fatalf("[%x, %x, %x] Error looking up ciphersuite: %s", vector.KEMID, vector.KDFID, vector.AEADID, err)
+	}
+
+	if vector.SKR != "" && vector.PKR != "" {
+		skRm, err := hex.DecodeString(vector.SKR)
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply hpke_test.go]: And for symmetry / brevity, you might just do encoding as well, and have `toHex / fromHex`.

File: hpke_test.go [hunk 27]
```diff
@@ -155,3 +159,347 @@ func TestModes(t *testing.T) {
...
+	// Sanity check
+	var decodedVector testVector
+	err = json.Unmarshal([]byte(encodedVectorString), &decodedVector)
+	if err != nil {
+		t.Fatalf("[%x, %x, %x] Error decoding test vector: %s", suite.KEM.ID(), suite.KDF.ID(), suite.AEAD.ID(), err)
+	}
+}
+
+func TestVectorGenerate(t *testing.T) {
+	if present := os.Getenv(testVectorKey); present == "" {
```
Comment (bifurcation) [inline hpke_test.go]: You could just do `len(os.Getenv(testVectorKey)) == 0`

File: README.md [hunk 28]
```diff
@@ -1,6 +1,13 @@
...
 
-https://tools.ietf.org/html/draft-barnes-cfrg-hpke
+https://tools.ietf.org/html/draft-irtf-cfrg-hpke
+
+## Test vector generation
+
+To generate test vectors, run:
+
+```
+$ HPKE_TEST_VECTORS=1 go test -v -run TestVectorGenerate
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply README.md]: Perfect!

File: crypto.go [hunk 29]
```diff
@@ -26,13 +26,17 @@ type dhScheme interface {
...
 	Unmarshal(enc []byte) (KEMPublicKey, error)
+	UnmarshalPrivate(enc []byte) (KEMPrivateKey, error)
 	DH(priv KEMPrivateKey, pub KEMPublicKey) ([]byte, error)
 	PublicKeySize() int
 }
 
 type dhkemScheme struct {
 	group dhScheme
+	skE   KEMPrivateKey
+	pkE   KEMPublicKey
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply crypto.go]: Good call!

File: hpke_test.go [hunk 30]
```diff
@@ -155,3 +159,347 @@ func TestModes(t *testing.T) {
...
+
+	var skR, skE KEMPrivateKey
+	var pkR, pkE KEMPublicKey
+
+	suite, err := AssembleCipherSuite(vector.KEMID, vector.KDFID, vector.AEADID)
+	if err != nil {
+		t.Fatalf("[%x, %x, %x] Error looking up ciphersuite: %s", vector.KEMID, vector.KDFID, vector.AEADID, err)
+	}
+
+	if vector.SKR != "" && vector.PKR != "" {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke_test.go]: Yep, good call. The `else` case we have now is leftover from the vector weaving logic.

File: hpke_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke_test.go]: Sounds good to me!

File: hpke_test.go [hunk 31]
```diff
@@ -155,3 +159,347 @@ func TestModes(t *testing.T) {
...
+			t.Fatalf("[%x, %x, %x] Error in Unmarshal: %s", vector.KEMID, vector.KDFID, vector.AEADID, err)
+		}
+	} else {
+		skR, pkR, err = suite.KEM.GenerateKeyPair(rand.Reader)
+		if err != nil {
+			t.Fatalf("[%x, %x, %x] Error generating initiator DH key pair: %s", vector.KEMID, vector.KDFID, vector.AEADID, err)
+		}
+	}
+
+	suite, err = assembleCipherSuiteWithEphemeralKeys(vector.KEMID, vector.KDFID, vector.AEADID, skE, pkE)
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke_test.go]: This one sets the ephemeral keying material, so it's needed.

File: crypto.go [hunk 32]
```diff
@@ -641,8 +754,32 @@ var aeads = map[AEADID]AEADScheme{
...
+		return dhkemScheme{group: ecdhScheme{curve: elliptic.P521()}, skE: sk, pkE: pk}, true
+	case KEM_SIKE503:
+		return sikeScheme{field: sidh.Fp503}, true
+	case KEM_SIKE751:
+		return sikeScheme{field: sidh.Fp751}, true
+	}
+	return nil, false
+}
+
+func assembleCipherSuiteWithEphemeralKeys(kemID KEMID, kdfID KDFID, aeadID AEADID, sk KEMPrivateKey, pk KEMPublicKey) (CipherSuite, error) {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply crypto.go]: 👍

File: hpke.go [hunk 33]
```diff
@@ -52,6 +54,10 @@ type AEADScheme interface {
 	NonceSize() int
 }
 
+type AEADContext interface {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke.go]: Yep, good call!

File: crypto.go [hunk 34]
```diff
@@ -47,22 +51,38 @@ func (s dhkemScheme) Marshal(pk KEMPublicKey) []byte {
...
+
+func (s dhkemScheme) getEphemeralKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey, error) {
+	if s.skE != nil && s.pkE != nil {
+		return s.skE, s.pkE, nil
+	} else {
+		return s.group.GenerateKeyPair(rand)
+	}
+}
+
+func (s dhkemScheme) Encap(rand io.Reader, pkR KEMPublicKey) ([]byte, []byte, KEMPrivateKey, error) {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply crypto.go]: Nope! We can generate them at the test vector entry point, shove them in the suite object, and then put them in the test vector.
